File: src/trading/strategies/stop-loss-arb/debug.ts
import {FloatCalculator as fc} from '../../../utils/float-calculator';
import {
    isHistoricalSnapshot,
    isHistoricalSnapshotsExhausted,
    isRandomSnapshot,
    restartRandomPrice,
} from '../../../utils/price-simulator';
import {OrderAction, Snapshot} from '../../brokerage-clients/brokerage-client';
import {StockState} from './types';

export function printPnLValues(stock: string, stockState: StockState): void {
    console.log(`stock: ${stock}`);
    console.log(`realizedPnL: ${stockState.realizedPnL}`);
    console.log(`exitPnL: ${stockState.exitPnL}`);
    console.log(`exitPnLAsPercentage: ${stockState.exitPnLAsPercentage}`);
    console.log(`maxMovingProfitAsPercentage: ${stockState.maxMovingProfitAsPercentage}`);
    console.log(`maxMovingLossAsPercentage: ${stockState.maxMovingLossAsPercentage}`);
    console.log('');
}

export async function debugRandomPrices(
    snapshot: Snapshot,
    stock: string,
    states: { [stock: string]: StockState },
    originalStates: { [stock: string]: StockState },
): Promise<void> {
    const stockState = states[stock];

    const aboveTopSell = fc.add(
        stockState.intervals[0][OrderAction.SELL].price,
        stockState.spaceBetweenIntervals,
    );
    if (fc.gte(snapshot.bid, aboveTopSell)) {
        debugUpperOrLowerBound('up', stock, states, originalStates);
        return;
    }

    const belowBottomBuy = fc.subtract(
        stockState.intervals[stockState.intervals.length - 1][OrderAction.BUY].price,
        stockState.spaceBetweenIntervals,
    );
    if (fc.lte(snapshot.ask, belowBottomBuy)) {
        debugUpperOrLowerBound('down', stock, states, originalStates);
    }
}

async function debugUpperOrLowerBound(
    upperOrLowerBound: 'up' | 'down',
    stock: string,
    states: { [stock: string]: StockState },
    originalStates: { [stock: string]: StockState },
): Promise<void> {
    if (states[stock].tradingLogs.length === 0) {
        restartRandomPrice();
        states[stock] = originalStates[stock];
        return;
    }

    if (
        upperOrLowerBound === 'up' &&
        states[stock].position < states[stock].targetPosition
    ) {
        debugger;
    } else if (
        upperOrLowerBound === 'down' &&
        states[stock].position > -states[stock].targetPosition
    ) {
        debugger;
    }

    restartRandomPrice();
    states[stock] = structuredClone(originalStates[stock]);
}

File: src/trading/strategies/stop-loss-arb/cpp/debug.cpp
#include "debug.hpp"

#include <format>

#include "price_simulator.hpp"

using namespace std;

void PrintPnLValues(const std::string stock, const StockState& stockState)
{
    Print(format("stock: {}", stock));
    Print(
        format("realizedPnLAsPercentage: {}", stockState.realizedPnLAsPercentage.str())
    );
    Print(format("exitPnLAsPercentage: {}", stockState.exitPnLAsPercentage.str()));
    Print(format(
        "maxMovingProfitAsPercentage: {}", stockState.maxMovingProfitAsPercentage.str()
    ));
    Print(format(
        "maxMovingLossAsPercentage: {}", stockState.maxMovingLossAsPercentage.str()
    ));
    Print("");
}

void DebugUpperOrLowerBound(
    const std::string& upperOrLowerBound,
    const std::string& stock,
    std::unordered_map<std::string, StockState>& states,
    const std::unordered_map<std::string, StockState>& originalStates
)
{
    if (upperOrLowerBound == "up" &&
        states[stock].position < states[stock].targetPosition)
    {
        Print("Place breakpoint here. No cross platform breakpoints in C++");
    }
    else if (upperOrLowerBound == "down" &&
             states[stock].position > -states[stock].targetPosition)
    {
        Print("Place breakpoint here. No cross platform breakpoints in C++");
    }

    RestartRandomPrice();
    states[stock] = originalStates.at(stock);

    return;
}

void DebugRandomPrices(
    const Snapshot& snapshot,
    const std::string& stock,
    std::unordered_map<std::string, StockState>& states,
    const std::unordered_map<std::string, StockState>& originalStates
)
{
    auto& stockState = states[stock];

    Decimal aboveTopSell =
        stockState.intervals[0].SELL.price + stockState.spaceBetweenIntervals;
    if (snapshot.bid >= aboveTopSell)
    {
        DebugUpperOrLowerBound("up", stock, states, originalStates);
        return;
    }

    Decimal belowBottomBuy =
        stockState.intervals.back().BUY.price - stockState.spaceBetweenIntervals;
    if (snapshot.ask <= belowBottomBuy)
    {
        DebugUpperOrLowerBound("down", stock, states, originalStates);
        return;
    }

    return;
}
